File: Hard/815. Bus Routes/815. Bus Routes.java
import java.util.*;

public class Solution {
    public int numBusesToDestination(int[][] routes, int source, int target) {
        if (source == target) {
            return 0;
        }

        Map<Integer, List<Integer>> stopToRoutes = new HashMap<>();
        Set<Integer> visitedRoutes = new HashSet<>();

        for (int i = 0; i < routes.length; i++) {
            for (int stop : routes[i]) {
                stopToRoutes.computeIfAbsent(stop, k -> new ArrayList<>()).add(i);
            }
        }

        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{source, 0});

        while (!queue.isEmpty()) {
            int[] currentStopAndSteps = queue.poll();
            int currentStop = currentStopAndSteps[0];
            int steps = currentStopAndSteps[1];

            for (int routeIndex : stopToRoutes.getOrDefault(currentStop, Collections.emptyList())) {
                if (visitedRoutes.contains(routeIndex)) {
                    continue;
                }

                visitedRoutes.add(routeIndex);

                for (int nextStop : routes[routeIndex]) {
                    if (nextStop == target) {
                        return steps + 1;
                    }

                    queue.offer(new int[]{nextStop, steps + 1});
                }
            }
        }

        return -1;
    }
}

File: Hard/815. Bus Routes/815. Bus Routes.rs
use std::collections::{HashMap, HashSet, VecDeque};

impl Solution {
    pub fn num_buses_to_destination(routes: Vec<Vec<i32>>, source: i32, target: i32) -> i32 {
        if source == target {
            return 0;
        }

        let mut stop_to_routes: HashMap<i32, Vec<usize>> = HashMap::new();
        for (i, route) in routes.iter().enumerate() {
            for stop in route {
                stop_to_routes
                    .entry(*stop)
                    .or_insert_with(Vec::new)
                    .push(i);
            }
        }

        let mut visited_routes = HashSet::new();
        let mut queue = VecDeque::new();
        queue.push_back((source, 0));

        while !queue.is_empty() {
            let (current_stop, steps) = queue.pop_front().unwrap();

            for &route_index in stop_to_routes.get(&current_stop).unwrap_or(&Vec::new()) {
                if visited_routes.contains(&route_index) {
                    continue;
                }
                visited_routes.insert(route_index);

                for &next_stop in routes[route_index].iter() {
                    if next_stop == target {
                        return steps + 1;
                    }
                    queue.push_back((next_stop, steps + 1));
                }
            }
        }

        -1
    }
}
